File: Facultate/Anul 2/Sem 2/MPP/Laburi/Lab5/JavaTema3/src/main/java/services/MainService.java
package services;

import models.Reservation;
import models.Ride;
import repos.EmployeeDBRepo;
import repos.ReservationsDBRepo;
import repos.RidesDBRepo;
import dtos.RideDTO;
import java.time.LocalDate;
import java.time.LocalTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import exceptions.ServiceException;
import utils.Observable;

public class MainService extends Observable {

    private final EmployeeDBRepo employeeDBRepo;
    private final RidesDBRepo ridesDBRepo;
    private final ReservationsDBRepo reservationsDBRepo;

    public MainService(EmployeeDBRepo employeeDBRepo, RidesDBRepo ridesDBRepo, ReservationsDBRepo reservationsDBRepo) {
        this.employeeDBRepo = employeeDBRepo;
        this.ridesDBRepo = ridesDBRepo;
        this.reservationsDBRepo = reservationsDBRepo;
    }


    public List<RideDTO> findRides(LocalDate rideDate, String departureHour, String destination) {
        List<Ride> rides = ridesDBRepo.findAll();
        Ride foundRide = rides.stream()
                .filter(ride -> ride.getDate().equals(rideDate) &&
                        ride.getDepartureTime().equals(LocalTime.parse(departureHour, DateTimeFormatter.ofPattern("HH:mm"))) &&
                        ride.getDestination().equals(destination))

                .findFirst()
                .orElse(null);
        if (foundRide == null) {
            throw new ServiceException("Ride not found");
        }
        List<Reservation> reservations = reservationsDBRepo.findReservationsByRideId(foundRide.getId());
        List<RideDTO> rideDTOList = new ArrayList<>();
        int seatNo = 1;
        for (Reservation reservation : reservations) {
            Integer seatsCount = reservation.getReservedSeatsCount();
            while (seatsCount > 0) {
                rideDTOList.add(new RideDTO(seatNo, reservation.getClientName()));
                seatNo++;
                seatsCount--;
            }
        }
        while (seatNo <= foundRide.getAvailableSeats()) {
            rideDTOList.add(new RideDTO(seatNo, "-"));
            seatNo++;
        }
        return rideDTOList;
    }

    public List<Ride> getAllRides() {
        return ridesDBRepo.findAll();
    }

    public void makeReservation(Long rideId, String text, int noSeats) {
        Reservation reservation = new Reservation(text, noSeats, rideId);
        reservationsDBRepo.save(reservation);
        Ride ride = ridesDBRepo.findOne(rideId).orElse(null);
        if(ride != null) {
            ride.setAvailableSeats(ride.getAvailableSeats() - noSeats);
            ridesDBRepo.update(ride);
        }
        notifyObservers();
    }

    public void cancelReservation(Long reservationId) {
        Reservation reservation = reservationsDBRepo.findOne(reservationId).orElse(null);
        if (reservation != null) {
            ridesDBRepo.findOne(reservation.getRideId()).ifPresent(ride -> {
                ride.setAvailableSeats(ride.getAvailableSeats() + reservation.getReservedSeatsCount());
                ridesDBRepo.update(ride);
            });
            reservationsDBRepo.delete(reservationId);
            notifyObservers();
        }
    }

    public Ride getRide(Long rideId) {
        return ridesDBRepo.findOne(rideId).orElse(null);
    }

    public List<Reservation> findReservations() {
        return reservationsDBRepo.findAll();
    }
}

File: Facultate/Anul 2/Sem 2/MPP/Laburi/CSharpTema3/CSharpLab3/services/MainService.cs

using CSharpLab3.dtos;
using CSharpLab3.exceptions;
using CSharpLab3.models;
using CSharpLab3.repos;
using CSharpLab3.utils;
using NLog;

namespace CSharpLab3.services;

public class MainService : Observable
{
    private readonly IEmployeeRepo _employeeRepo;
    private readonly IRidesRepo _ridesRepo;
    private readonly IReservationsRepo _reservationsRepo;
    private static readonly Logger Logger = LogManager.GetCurrentClassLogger();

    public MainService(IEmployeeRepo employeeRepo, IRidesRepo ridesRepo, IReservationsRepo reservationsRepo)
    {
        _employeeRepo = employeeRepo;
        _ridesRepo = ridesRepo;
        _reservationsRepo = reservationsRepo;
    }

    public List<RideDTO> FindRides(DateOnly rideDate, string departureHour, string destination)
    {
        var departureTime = TimeOnly.Parse(departureHour);
        var foundRide = _ridesRepo.FindAll()
            .FirstOrDefault(r =>
                r.Date == rideDate &&
                r.DepartureTime == departureTime &&
                r.Destination == destination)
            ?? throw new ServiceException("Ride not found");

        var reservations = _reservationsRepo.FindReservationsByRide(foundRide.Id);        
        var result = new List<RideDTO>();
        int seatNo = 1;

        foreach (var reservation in reservations)
        {
            for (int i = 0; i < reservation.ReservedSeatsCount; i++)
                result.Add(new RideDTO(seatNo++, reservation.ClientName));
        }

        while (seatNo <= foundRide.AvailableSeats)
            result.Add(new RideDTO(seatNo++, "-"));

        return result;
    }

    public List<Ride> GetAllRides() => _ridesRepo.FindAll().ToList();

    public void MakeReservation(long rideId, string clientName, int noSeats)
    {
        var reservation = new Reservation
        {
            ClientName = clientName,
            ReservedSeatsCount = noSeats,
            RideId = rideId
        };
        _reservationsRepo.Save(reservation);

        var ride = _ridesRepo.FindOne(rideId);
        if (ride != null)
        {
            ride.AvailableSeats -= noSeats;
            _ridesRepo.Update(ride);
        }

        NotifyObservers();
        Logger.Info($"Reservation made: {clientName}, {noSeats} seats, ride {rideId}");
    }

    public void CancelReservation(long reservationId)
    {
        var reservation = _reservationsRepo.FindOne(reservationId);
        if (reservation == null) return;

        var ride = _ridesRepo.FindOne(reservation.RideId);
        if (ride != null)
        {
            ride.AvailableSeats += reservation.ReservedSeatsCount;
            _ridesRepo.Update(ride);
        }

        _reservationsRepo.Delete(reservationId);
        NotifyObservers();
        Logger.Info($"Reservation {reservationId} cancelled");
    }

    public Ride? GetRide(long rideId) => _ridesRepo.FindOne(rideId);

    public List<Reservation> FindReservations() => _reservationsRepo.FindAll().ToList();
}